### PR TITLE
Tic tac toe

### PR DESCRIPTION
**Pull request by Nosoyleon**
## Summary
1. Taking turns fo 'O' and 'X'
2. Geting the Winner
3. Lifting Up state 
4. Adding Time Travel
5. Improving rendering of squares
6. A little bit of styling
7. PropTypes checking

## Screenshots
image
nov-16-2018 16-03-11

## Trello Card
https://trello.com/c/j423o2Y6/43-react-app-tic-tac-toe
https://trello.com/c/0hz98uXM/44-proptypes-checking

### CODE REVIEW

File: react/training-project/src/app/components/Board/index.js [hunk 1]
```diff
@@ -1,39 +1,26 @@
...
-    squares: Array(9).fill(null)
-  };
-  handleClick = i => {
-    const squares = this.state.squares.slice();
-    squares[i] = 'X';
-    this.setState({ squares });
-  };
-
-  renderSquare = i => <Square value={this.state.squares[i]} key={i} id={i} handleClick={this.handleClick} />;
+  renderSquare = i => (
```
Comment (BJordi) [inline react/training-project/src/app/components/Board/index.js]: Why did you add the parenthesis? :thinking:

File: react/training-project/src/app/components/Board/index.js [hunk 2]
```diff
@@ -1,39 +1,26 @@
...
-        <div className={style.boardRow}>
-          {Array.from(Array(NUMBER_OF_COLUMNS).keys()).map(i => this.renderSquare(i))}
-        </div>
-        <div className={style.boardRow}>
-          {Array.from(Array(NUMBER_OF_COLUMNS).keys()).map(i => this.renderSquare(i + 3))}
-        </div>
-        <div className={style.boardRow}>
-          {Array.from(Array(NUMBER_OF_COLUMNS).keys()).map(i => this.renderSquare(i + 6))}
-        </div>
+        {Array.from(Array(NUMBER_OF_SQUARES).keys()).map(i => this.renderSquare(i))}
```
Comment (BJordi) [inline react/training-project/src/app/components/Board/index.js]: `i` -> `index`

File: react/training-project/src/app/components/Game/index.js [hunk 3]
```diff
@@ -1,21 +1,75 @@
...
-        <ol>{/* TODO */}</ol>
+class Game extends Component {
+  state = {
+    history: [{ squares: Array(9).fill(null) }],
+    stepNumber: 0,
+    xIsNext: true
+  };
+
+  calculateWinner = squares => {
+    const lines = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]];
```
Comment (BJordi) [inline react/training-project/src/app/components/Game/index.js]: I would extract it or you'll be creating this array every time you call `calculateWinner`.

File: react/training-project/src/app/components/Game/index.js [hunk 4]
```diff
@@ -1,21 +1,75 @@
...
+  render() {
+    const history = this.state.history.slice();
+    const current = history[this.state.stepNumber];
+    const winner = this.calculateWinner(current.squares);
+
+    const moves = history.map((step, move) => {
+      const desc = move ? `Go to move #${move}` : `Go to game start`;
+      return (
+        <li key={move.toString()}>
+          <button onClick={() => this.jumpTo(move)}>{desc}</button>
```
Comment (BJordi) [inline react/training-project/src/app/components/Game/index.js]: Extract `() => this.jumpTo(move)`

File: react/training-project/src/app/components/Game/index.js [hunk 5]
```diff
@@ -1,21 +1,75 @@
...
+    for (let i = 0; i < lines.length; i += 1) {
+      const [a, b, c] = lines[i];
+      if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
+        return squares[a];
+      }
+    }
+    return null;
+  };
+
+  handleClick = i => {
```
Comment (BJordi) [inline react/training-project/src/app/components/Game/index.js]: Use a more descriptive name.

File: react/training-project/src/app/components/Game/index.js [hunk 6]
```diff
@@ -1,21 +1,75 @@
...
+    const history = this.state.history.slice(0, this.state.stepNumber + 1);
```
Comment (BJordi) [inline react/training-project/src/app/components/Game/index.js:29]: Don't use literals.

File: react/training-project/src/app/components/Game/index.js [hunk 7]
```diff
@@ -1,21 +1,75 @@
...
+    const current = history[history.length - 1];
+    const squares = current.squares.slice();
+    if (this.calculateWinner(squares) || squares[i]) return;
+    squares[i] = this.state.xIsNext ? 'X' : 'O';
```
Comment (BJordi) [inline react/training-project/src/app/components/Game/index.js]: Same here. Extract them into constants.

**Review by BJordi — CHANGES_REQUESTED** [review]

File: react/training-project/src/app/components/Board/index.js (hunk 1 above)
Comment (Nosoyleon) [reply react/training-project/src/app/components/Board/index.js]: I was trying to make happy the linter :(

File: react/training-project/src/app/components/App/index.js [hunk 8]
```diff
@@ -11,7 +11,7 @@ class App extends Component {
       <div className={style.app}>
         <header className={style.appHeader}>
           <img src={logo} className={style.appLogo} alt="logo" />
-          <h1 className={style.appTitle}>Welcome to React</h1>
+          <h1 className={style.appTitle}>TicTacToe</h1>
```
Comment (BJordi) [inline react/training-project/src/app/components/App/index.js]: Move it to i18n. If you're not using it yet, then, move it into a constant.

File: react/training-project/src/app/components/Button/index.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+function Button(props) {
```
Comment (BJordi) [inline react/training-project/src/app/components/Button/index.js]: Turn it into a class component. Remember we use smar and /dumb components. For smart ones, we use class components, for dumb ones we use functions.

Comment (BJordi) [inline react/training-project/src/app/components/Button/index.js]: For future components, you can do this:
```
function Button({move, children}) {
...
// Use them like {children} or {move} instead of {props.children} or {props.move}
}
```

File: react/training-project/src/app/components/Game/index.js [hunk 10]
```diff
@@ -1,21 +1,79 @@
...
+  state = {
+    history: [{ squares: Array(9).fill(null) }],
+    stepNumber: 0,
+    xIsNext: true
+  };
+
+  LINES_TO_CHECK = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]];
+
+  calculateWinner = squares => {
+    for (let i = 0; i < this.LINES_TO_CHECK.length; i += 1) {
```
Comment (BJordi) [inline react/training-project/src/app/components/Game/index.js]: `i++` or `++i`

**Review by BJordi — CHANGES_REQUESTED** [review]

File: react/training-project/src/app/components/Board/index.js [hunk 11]
```diff
@@ -1,47 +1,30 @@
...
-          {Array.from(Array(NUMBER_OF_COLUMNS).keys()).map(i => this.renderSquare(i))}
-        </div>
-        <div className={style.boardRow}>
-          {Array.from(Array(NUMBER_OF_COLUMNS).keys()).map(i => this.renderSquare(i + 3))}
-        </div>
-        <div className={style.boardRow}>
-          {Array.from(Array(NUMBER_OF_COLUMNS).keys()).map(i => this.renderSquare(i + 6))}
-        </div>
-      </React.Fragment>
+    const NUMBER_OF_SQUARES = 9;
```
Comment (BJordi) [inline react/training-project/src/app/components/Board/index.js]: I think I would move it out of the render or this constant will be created again in every render.

**Review by BJordi — APPROVED** [review]
